File: yogi-dotnet/yogi/Context.cs
/*
 * This file is part of the Yogi distribution https://github.com/yohummus/yogi.
 * Copyright (c) 2018 Johannes Bergmann.
 *
 * This library is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library. If not, see <http://www.gnu.org/licenses/>.
 */

using System;
using System.Runtime.InteropServices;

public static partial class Yogi
{
    partial class Api
    {
        // === YOGI_ContextCreate ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ContextCreateDelegate(ref IntPtr context);

        public static ContextCreateDelegate YOGI_ContextCreate
            = Library.GetDelegateForFunction<ContextCreateDelegate>(
                "YOGI_ContextCreate");

        // === YOGI_ContextPoll ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ContextPollDelegate(SafeObjectHandle context, ref int count);

        public static ContextPollDelegate YOGI_ContextPoll
            = Library.GetDelegateForFunction<ContextPollDelegate>(
                "YOGI_ContextPoll");

        // === YOGI_ContextPollOne ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ContextPollOneDelegate(SafeObjectHandle context, ref int count);

        public static ContextPollOneDelegate YOGI_ContextPollOne
            = Library.GetDelegateForFunction<ContextPollOneDelegate>(
                "YOGI_ContextPollOne");

        // === YOGI_ContextRun ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ContextRunDelegate(SafeObjectHandle context, ref int count,
            long duration);

        public static ContextRunDelegate YOGI_ContextRun
            = Library.GetDelegateForFunction<ContextRunDelegate>(
                "YOGI_ContextRun");

        // === YOGI_ContextRunOne ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ContextRunOneDelegate(SafeObjectHandle context, ref int count,
            long duration);

        public static ContextRunOneDelegate YOGI_ContextRunOne
            = Library.GetDelegateForFunction<ContextRunOneDelegate>(
                "YOGI_ContextRunOne");

        // === YOGI_ContextRunInBackground ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ContextRunInBackgroundDelegate(SafeObjectHandle context);

        public static ContextRunInBackgroundDelegate YOGI_ContextRunInBackground
            = Library.GetDelegateForFunction<ContextRunInBackgroundDelegate>(
                "YOGI_ContextRunInBackground");

        // === YOGI_ContextStop ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ContextStopDelegate(SafeObjectHandle context);

        public static ContextStopDelegate YOGI_ContextStop
            = Library.GetDelegateForFunction<ContextStopDelegate>(
                "YOGI_ContextStop");

        // === YOGI_ContextWaitForRunning ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ContextWaitForRunningDelegate(SafeObjectHandle context, long duration);

        public static ContextWaitForRunningDelegate YOGI_ContextWaitForRunning
            = Library.GetDelegateForFunction<ContextWaitForRunningDelegate>(
                "YOGI_ContextWaitForRunning");

        // === YOGI_ContextWaitForStopped ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ContextWaitForStoppedDelegate(SafeObjectHandle context, long duration);

        public static ContextWaitForStoppedDelegate YOGI_ContextWaitForStopped
            = Library.GetDelegateForFunction<ContextWaitForStoppedDelegate>(
                "YOGI_ContextWaitForStopped");

        // === YOGI_ContextPost ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate void ContextPostFnDelegate(IntPtr userarg);

        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ContextPostDelegate(SafeObjectHandle context,
            ContextPostFnDelegate fn, IntPtr userarg);

        public static ContextPostDelegate YOGI_ContextPost
            = Library.GetDelegateForFunction<ContextPostDelegate>(
                "YOGI_ContextPost");
    }

    /// <summary>
    /// Scheduler for the execution of asynchronous operations.
    ///
    /// Once an asynchronous operation finishes, the corresponding handler function
    /// is added to the context's event queue and executed through once of the
    /// Poll... or Run... functions.
    /// </summary>
    public class Context : Object
    {
        /// <summary>
        /// Constructor.
        /// </summary>
        public Context()
        : base(Create())
        {
        }

        /// <summary>
        /// Runs the context's event processing loop to execute ready handlers.
        ///
        /// This function runs handlers (internal and user-supplied such as functions
        /// registered through Post()) that are ready to run, without blocking, until
        /// the Stop() function has been called or there are no more ready handlers.
        ///
        /// This function must be called from outside any handler functions that are
        /// being executed through the context.
        /// </summary>
        /// <returns>Number of executed handlers.</returns>
        public int Poll()
        {
            int count = -1;
            int res = Api.YOGI_ContextPoll(Handle, ref count);
            CheckErrorCode(res);
            return count;
        }

        /// <summary>
        /// Runs the context's event processing loop to execute at most one ready handler.
        ///
        /// This function runs at most one handler (internal and user-supplied such as
        /// functions registered through Post()) that are ready to run, without blocking.
        ///
        /// This function must be called from outside any handler functions that are being
        /// executed through the context.
        /// </summary>
        /// <returns>Number of executed handlers.</returns>
        public int PollOne()
        {
            int count = -1;
            int res = Api.YOGI_ContextPollOne(Handle, ref count);
            CheckErrorCode(res);
            return count;
        }

        /// <summary>
        /// Runs the context's event processing loop for the specified duration.
        ///
        /// This function blocks while running the context's event processing loop and
        /// calling dispatched handlers (internal and user-supplied such as functions
        /// registered through Post()) for the specified duration unless Stop() is called
        /// within that time.
        ///
        /// This function must be called from outside any handler functions that are being
        /// executed through the context.
        /// </summary>
        /// <param name="duration">Duration (infinity by default).</param>
        /// <returns>Number of executed handlers.</returns>
        public int Run([Optional] Duration duration)
        {
            long dur = duration == null ? -1 : DurationToApiDuration(duration);

            int count = -1;
            int res = Api.YOGI_ContextRun(Handle, ref count, dur);
            CheckErrorCode(res);
            return count;
        }

        /// <summary>
        /// Runs the context's event processing loop for the specified duration.
        ///
        /// This function blocks while running the context's event processing loop and
        /// calling dispatched handlers (internal and user-supplied such as functions
        /// registered through Post()) for the specified duration unless Stop() is called
        /// within that time.
        ///
        /// This function must be called from outside any handler functions that are being
        /// executed through the context.
        /// </summary>
        /// <param name="duration">Duration.</param>
        /// <returns>Number of executed handlers.</returns>
        public int Run(TimeSpan duration)
        {
            return Run(new Duration(duration));
        }

        /// <summary>
        /// Runs the context's event processing loop for the specified duration to execute
        /// at most one handler.
        ///
        /// This function blocks while running the context's event processing loop and
        /// calling dispatched handlers (internal and user-supplied such as functions
        /// registered through Post()) for the specified duration until a single handler
        /// function has been executed, unless Stop() is called within that time.
        ///
        /// This function must be called from outside any handler functions that are being
        /// executed through the context.
        /// </summary>
        /// <param name="duration">Duration (infinity by default).</param>
        /// <returns>Number of executed handlers.</returns>
        public int RunOne([Optional] Duration duration)
        {
            long dur = duration == null ? -1 : DurationToApiDuration(duration);

            int count = -1;
            int res = Api.YOGI_ContextRunOne(Handle, ref count, dur);
            CheckErrorCode(res);
            return count;
        }

        /// <summary>
        /// Runs the context's event processing loop for the specified duration to execute
        /// at most one handler.
        ///
        /// This function blocks while running the context's event processing loop and
        /// calling dispatched handlers (internal and user-supplied such as functions
        /// registered through Post()) for the specified duration until a single handler
        /// function has been executed, unless Stop() is called within that time.
        ///
        /// This function must be called from outside any handler functions that are being
        /// executed through the context.
        /// </summary>
        /// <param name="duration">Duration.</param>
        /// <returns>Number of executed handlers.</returns>
        public int RunOne(TimeSpan duration)
        {
            return RunOne(new Duration(duration));
        }

        /// <summary>
        /// Starts an internal thread for running the context's event processing loop.
        ///
        /// This function starts a threads that runs the context's event processing loop
        /// in the background. It relieves the user from having to start a thread and
        /// calling the appropriate Run... or Poll... functions themself. The thread can
        /// be stopped using Stop().
        ///
        /// This function must be called from outside any handler functions that are being
        /// executed through the context.
        /// </summary>
        public void RunInBackground()
        {
            int res = Api.YOGI_ContextRunInBackground(Handle);
            CheckErrorCode(res);
        }

        /// <summary>
        /// Stops the context's event processing loop.
        ///
        /// This function signals the context to stop running its event processing loop.
        /// This causes Run... functions to return as soon as possible and it terminates
        /// the thread started via RunInBackground().
        /// </summary>
        public void Stop()
        {
            int res = Api.YOGI_ContextStop(Handle);
            CheckErrorCode(res);
        }

        /// <summary>
        /// Blocks until the context's event processing loop is being run or until the
        /// specified timeout is reached.
        ///
        /// This function must be called from outside any handler functions that are being
        /// executed through the context.
        /// </summary>
        /// <param name="duration">Maximum time to wait (infinity by default).</param>
        /// <returns>True if the context's event processing loop is running within the
        /// specified duration and false otherwise.</returns>
        public bool WaitForRunning([Optional] Duration duration)
        {
            long dur = duration == null ? -1 : DurationToApiDuration(duration);

            int res = Api.YOGI_ContextWaitForRunning(Handle, dur);
            if ((ErrorCode)res == ErrorCode.Timeout) return false;
            CheckErrorCode(res);
            return true;
        }

        /// <summary>
        /// Blocks until the context's event processing loop is being run or until the
        /// specified timeout is reached.
        ///
        /// This function must be called from outside any handler functions that are being
        /// executed through the context.
        /// </summary>
        /// <param name="duration">Maximum time to wait.</param>
        /// <returns>True if the context's event processing loop is running within the
        /// specified duration and false otherwise.</returns>
        public bool WaitForRunning(TimeSpan duration)
        {
            return WaitForRunning(new Duration(duration));
        }

        /// <summary>
        /// Blocks until no thread is running the context's event processing
        /// loop or until the specified timeout is reached.
        ///
        /// This function must be called from outside any handler functions that are being
        /// executed through the context.
        /// </summary>
        /// <param name="duration">Maximum time to wait (infinity by default).</param>
        /// <returns>True if the context's event processing loop is not running within the
        /// specified duration and false otherwise.</returns>
        public bool WaitForStopped([Optional] Duration duration)
        {
            long dur = duration == null ? -1 : DurationToApiDuration(duration);

            int res = Api.YOGI_ContextWaitForStopped(Handle, dur);
            if ((ErrorCode)res == ErrorCode.Timeout) return false;
            CheckErrorCode(res);
            return true;
        }

        /// <summary>
        /// Blocks until no thread is running the context's event processing
        /// loop or until the specified timeout is reached.
        ///
        /// This function must be called from outside any handler functions that are being
        /// executed through the context.
        /// </summary>
        /// <param name="duration">Maximum time to wait.</param>
        /// <returns>True if the context's event processing loop is not running within the
        /// specified duration and false otherwise.</returns>
        public bool WaitForStopped(TimeSpan duration)
        {
            return WaitForStopped(new Duration(duration));
        }

        /// <summary>
        /// Adds the given function to the context's event processing queue to
        /// be executed and returns immediately.
        ///
        /// The handler fn will only be executed after this function returns and
        /// only by a thread running the context's event processing loop.
        /// </summary>
        /// <param name="fn">The function to call from within the context.</param>
        public void Post(Action fn)
        {
            Api.ContextPostFnDelegate wrapper = (userarg) => {
                try
                {
                    fn();
                }
                finally
                {
                    GCHandle.FromIntPtr(userarg).Free();
                }
            };
            var wrapperHandle = GCHandle.Alloc(wrapper);

            try
            {
                var wrapperPtr = GCHandle.ToIntPtr(wrapperHandle);
                int res = Api.YOGI_ContextPost(Handle, wrapper, wrapperPtr);
                CheckErrorCode(res);
            }
            catch
            {
                wrapperHandle.Free();
                throw;
            }

        }

        static IntPtr Create()
        {
            var handle = new IntPtr();
            int res = Api.YOGI_ContextCreate(ref handle);
            CheckErrorCode(res);
            return handle;
        }
    }
}

File: yogi-core/src/objects/context.cc
/*
 * This file is part of the Yogi distribution https://github.com/yohummus/yogi.
 * Copyright (c) 2019 Johannes Bergmann.
 *
 * This library is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library. If not, see <http://www.gnu.org/licenses/>.
 */

#include "context.h"
#include "../api/errors.h"

#include <boost/asio/post.hpp>
#include <signal.h>

YOGI_DEFINE_INTERNAL_LOGGER("Context")

namespace objects {

Context::Context() : ioc_(1), work_(ioc_), signals_(ioc_), running_(false) {
  SetLoggingPrefix(*this);
}

Context::~Context() {
  Stop();
  WaitForStopped(std::chrono::nanoseconds::max());
}

int Context::Poll() {
  return RunImpl([&] { return ioc_.poll(); });
}

int Context::PollOne() {
  return RunImpl([&] { return ioc_.poll_one(); });
}

int Context::Run(std::chrono::nanoseconds dur) {
  return RunImpl([&] {
    if (dur == dur.max()) {
      return ioc_.run();
    } else {
      return ioc_.run_for(dur);
    }
  });
}

int Context::RunOne(std::chrono::nanoseconds dur) {
  return RunImpl([&] {
    if (dur == dur.max()) {
      return ioc_.run_one();
    } else {
      return ioc_.run_one_for(dur);
    }
  });
}

void Context::RunInBackground() {
  SetRunningFlagAndReset();
  thread_ = std::thread([&] {
    try {
      ioc_.run();
    } catch (const std::exception& e) {
      LOG_FAT("Exception caught in context background thread: " << e.what());
    } catch (...) {
      LOG_FAT("Unknown Exception caught in context background thread");
    }

    ClearRunningFlag();
  });
}

void Context::Stop() {
  std::lock_guard<std::mutex> lock(mutex_);
  ioc_.stop();
}

bool Context::WaitForRunning(std::chrono::nanoseconds timeout) {
  std::unique_lock<std::mutex> lock(mutex_);
  bool timed_out = false;
  if (timeout == timeout.max()) {
    cv_.wait(lock, [&] { return running_; });
  } else {
    timed_out = !cv_.wait_for(lock, timeout, [&] { return running_; });
  }

  return !timed_out;
}

bool Context::WaitForStopped(std::chrono::nanoseconds timeout) {
  std::unique_lock<std::mutex> lock(mutex_);
  bool timed_out = false;
  if (timeout == timeout.max()) {
    cv_.wait(lock, [&] { return !running_; });
  } else {
    timed_out = !cv_.wait_for(lock, timeout, [&] { return !running_; });
  }

  if (!timed_out && thread_.joinable()) {
    thread_.join();
  }

  return !timed_out;
}

void Context::Post(std::function<void()> fn) { boost::asio::post(ioc_, fn); }

void Context::AwaitSignal(api::Signals sigs, SignalHandler signal_handler) {
  std::lock_guard<std::mutex> lock(mutex_);

  boost::system::error_code ec;
  signals_.cancel(ec);
  if (ec) {
    LOG_ERR("Could not cancel wait for signal operation: " << ec.message());
    throw api::Error(YOGI_ERR_UNKNOWN);
  }

  signals_.clear(ec);
  if (ec) {
    LOG_ERR("Could not clear signal set: " << ec.message());
    throw api::Error(YOGI_ERR_UNKNOWN);
  }

  if (sigs & api::kSigInt) {
    signals_.add(SIGINT, ec);
  }

  if (!ec && sigs & api::kSigTerm) {
    signals_.add(SIGTERM, ec);
  }

  if (ec) {
    LOG_ERR("Could not add SIGINT to signal set: " << ec.message());
    throw api::Error(YOGI_ERR_UNKNOWN);
  }

  signals_.async_wait([=](auto& ec, int sig_num) {
    auto res = api::kSuccess;
    auto sig = sigs;

    if (!ec) {
      switch (sig_num) {
        case SIGINT:
          sig = api::kSigInt;
          break;

        case SIGTERM:
          sig = api::kSigTerm;
          break;

        default:
          YOGI_NEVER_REACHED;
      }
    } else if (ec == boost::asio::error::operation_aborted) {
      res = api::Error(YOGI_ERR_CANCELED);
    } else {
      LOG_ERR("Could not wait for signal set: " << ec.message());
      res = api::Error(YOGI_ERR_UNKNOWN);
    }

    signal_handler(res, sig);
  });
}

void Context::CancelAwaitSignal() { signals_.cancel(); }

void Context::SetRunningFlagAndReset() {
  std::lock_guard<std::mutex> lock(mutex_);
  if (running_) {
    throw api::Error(YOGI_ERR_BUSY);
  }

  running_ = true;
  ioc_.restart();
  cv_.notify_all();
}

void Context::ClearRunningFlag() {
  std::lock_guard<std::mutex> lock(mutex_);
  YOGI_ASSERT(running_ == true);
  running_ = false;
  cv_.notify_all();
}

template <typename Fn>
int Context::RunImpl(Fn fn) {
  SetRunningFlagAndReset();
  int cnt = static_cast<int>(fn());
  ClearRunningFlag();
  return cnt;
}

}  // namespace objects
